Update background task documentation for CMS 17.5 (and 18.0) (#8082)
umbraco/UmbracoDocs: 17/umbraco-cms/extend-your-project/server-side-extensions/scheduling.md

# Scheduling

You can run recurring code using a recurring background job. The recommended way is to inherit from the abstract `RecurringBackgroundJobBase` class. The class provides defaults for delay, server roles, and event handling, so you only need to set the `Period` and implement `RunJobAsync(CancellationToken)`.

> If you need more control, you can implement the `IRecurringBackgroundJob` interface directly.

Reviewer: The interface doesn't really give you more control, but provides default values that you can override if needed.

Author: Have rewritten.

> Once you have created your background job class, register it using a Composer. The job is detected at startup and a new `HostedService` is created to run it.

Reviewer: The more important part is what extension method to use, not that you need to do it using a composer:

```suggestion
Once you have created your background job class, register it using `AddRecurringBackgroundJob<TJob>()`. The job is detected at startup and a new hosted service is created to run it.
```

{% hint style="warning" %}
Be aware you may or may not want this background job to run on all servers. If you are using Load Balancing with multiple servers, see the [load balancing documentation](../../run-in-production/infrastructure-and-ops/server-setup/load-balancing/) for more information.
{% endhint %}

## `RecurringBackgroundJobBase` Properties and Methods

The members below are defined on `IRecurringBackgroundJob` and either inherited or overridden via `RecurringBackgroundJobBase`. The signatures and behavior are the same whether you inherit the base class or implement the interface directly.

### Period

Defines how often the job runs. This property is a `TimeSpan`.

```csharp
// Run this job every 5 minutes
public override TimeSpan Period => TimeSpan.FromMinutes(5);
```

> Set `Period` to `Timeout.InfiniteTimeSpan` to disable automatic scheduling. The job then only runs when [triggered manually](scheduling.md#on-demand-triggering).

Reviewer: ```suggestion
Set `Period` to `Timeout.InfiniteTimeSpan` to disable recurring runs. The job then only runs when [triggered manually](scheduling.md#on-demand-triggering).
```

### Delay

> Defines how long to wait after application startup before running the job for the first time. The default is 3 minutes.

Reviewer: Might be good to explain why you'd want to wait, e.g. to ensure you're not doing additional resource intensive work during startup, give caches time to populate, etc.

```csharp
// Wait 1 minute after application startup before running this job for the first time.
public override TimeSpan Delay => TimeSpan.FromMinutes(1);
```

> Set `Delay` to `Timeout.InfiniteTimeSpan` to skip the automatic first run. The job then only runs when triggered manually.

Reviewer: This doesn't really skip the first run, but waits for a manual trigger before the recurring runs actually start, which is a slightly different result:

```suggestion
Set `Delay` to `Timeout.InfiniteTimeSpan` to not automatically start the recurring runs. The job then only runs (and starts recurring runs) when triggered manually.
```

        _scopeProvider = scopeProvider;
    }

    public override TimeSpan Period => TimeSpan.FromMinutes(60);

    // Run on all servers
    public override ServerRole[] ServerRoles => Enum.GetValues<ServerRole>();

> 

Reviewer: Because this is a database-only operation, it doesn't make sense to run it on all servers:

```suggestion
```

### Registering with a composer

> Create a composer and register the background job with `AddRecurringBackgroundJob`.

Reviewer: ```suggestion
Create a composer and register the background job with `AddRecurringBackgroundJob<TJob>()`.
```

| Overload | Behavior after the triggered execution |
| --- | --- |
| `TriggerExecution()` | Resume the original schedule. If the triggered execution overshot the next scheduled tick, skip it to avoid double-execution. Same as `NextExecutionStrategy.None`. |
| `TriggerExecution(NextExecutionStrategy.Reset)` | Wait a full period. The schedule shifts forward from the triggered execution. |
| `TriggerExecution(NextExecutionStrategy.Replace)` | The triggered execution replaces the next scheduled tick. The following execution occurs one full period after the originally-scheduled time. |
| `TriggerExecution(TimeSpan nextDelay)` | Wait the specified custom delay before the next execution. |

For example, to trigger a job and reset its schedule:

```csharp
_trigger.TriggerExecution(NextExecutionStrategy.Reset);
```

### Manual-only jobs

To create a job that only runs when triggered manually, set `Period` to `Timeout.InfiniteTimeSpan`. The job is registered and a hosted service is created for it, but no automatic execution occurs.

```csharp
public override TimeSpan Period => Timeout.InfiniteTimeSpan;
```

> Combine this with `Delay => Timeout.InfiniteTimeSpan` to also skip the initial run after startup.

Reviewer: ```suggestion
To create a job that only runs when triggered manually, set both `Period` and `Delay` to `Timeout.InfiniteTimeSpan`. The job is registered and a hosted service is created for it, but no automatic execution occurs.

```csharp
public override TimeSpan Period => Timeout.InfiniteTimeSpan;

public override TimeSpan Deplay => Timeout.InfiniteTimeSpan;
```

`RecurringHostedServiceBase` is a low-level base class. It inherits from .NET's `BackgroundService` and runs the job on a recurring basis using a cancellable wait loop. The loop supports periodic execution, manual triggering, exception resilience, and cooperative cancellation on host shutdown.

`RecurringBackgroundJobHostedService` is an Umbraco-specific hosted service that extends `RecurringHostedServiceBase`. It uses Umbraco system services to ensure that your jobs only execute once Umbraco is up and running. It checks:

* Server Roles - see above for more discussion about Server roles.
* MainDom - The `MainDom` lock ensures that only one instance of Umbraco is running at a time on a given machine. This ensures the integrity of certain files used by Umbraco. See [Host Synchronization](../../run-in-production/infrastructure-and-ops/server-setup/load-balancing/azure-web-apps.md#host-synchronization) for more details.
* Runtime State - On a fresh install or when waiting for a database upgrade, Umbraco may not be fully up and running yet.

When any of these checks fail, the execution is ignored and the loop waits for `IgnoredDelay` before re-evaluating.

## Notifications

> The `RecurringBackgroundJobHostedService` publishes a number of notifications that can be hooked to report on the status of background jobs. All notifications extend from the base `Umbraco.Cms.Infrastructure.Notifications.RecurringBackgroundJobNotification` class.

Reviewer: ```suggestion
The `RecurringBackgroundJobHostedService` publishes a number of notifications that can be used to report on the status of background jobs. All notifications extend from the base `Umbraco.Cms.Infrastructure.Notifications.RecurringBackgroundJobNotification` class.
```

* Executing
* Executed
* Failed
* Canceled

> * Ignored

Reviewer: This list is missing some notifications, don't use the full type name and have no short descriptions (so you have to read the next paragraphs to quickly understand their use):

```suggestion
* `RecurringBackgroundJobStartingNotification` - published before starting the recurring job
* `RecurringBackgroundJobStartedNotification` - published after the recurring job has started
* `RecurringBackgroundJobExecutingNotification` - published before running the job
* `RecurringBackgroundJobIgnoredNotification` - published when the job is ignored (see `IgnoredDelay`)
* `RecurringBackgroundJobExecutedNotification` - published after `RunJobAsync()` is called
* `RecurringBackgroundJobCanceledNotification` - published when the job was cancelled due to application shutdown
* `RecurringBackgroundJobFailedNotification` - published when an unhandled exception was thrown while running the job
* `RecurringBackgroundJobStoppingNotification` - published before stopping the recurring job
* `RecurringBackgroundJobStoppedNotification` - published after the recurring job has stopped
```

Quick note: `RecurringBackgroundJobNotification` isn't marked as abstract, which might be something we should fix in an upcoming major.

Author: I've added a `TODO` in the code base to handle in the next major.

### Start/Stop

The Starting/Started and Stopping/Stopped notification pairs are published when the `RecurringBackgroundJobHostedService` is started or stopped. The start event normally occurs soon after application start as part of the .NET WebHost startup process. Similarly, the stop event happens as part of application shutdown.

These notifications are there to support low-level debugging of background jobs to ensure they are starting and stopping correctly. Due to the timing of the notification, all handlers associated with these notifications should not depend on any Umbraco services, including database access.

### Ignored

The Ignored notification is published when a background job's schedule is triggered, but the Umbraco runtime checks prevent it from running.

This notification is there to support low-level debugging of background jobs to ascertain why they are or aren't running. As the runtime checks include runtime state readiness, this event may be triggered during the install phase. Any notification handlers associated with this notification should also conduct their own checks before relying on Umbraco services, including database access.

> 

Reviewer: ```suggestion

For **ignored** job runs, the following notifications are published:

1. Executing
2. Ignored
```
